Web resource 4

otaviojacobi opened the pull request.

fisehara commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,105 @@
+import { Upload } from '@aws-sdk/lib-storage';
+import { v4 as uuidv4 } from 'uuid';
+
+export interface S3HandlerConfig {
+	bucket?: string;
+	region?: string;
+	accessKeyId?: string;
+	secretAccessKey?: string;
+	endpoint?: string;
+}
```
Are all properties actually optional? Or does it not make sense to have any property left optional? 
We may find issues in the usage on the typings rather than at run time when the implementation doesn't provide mandatory fields

fisehara commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,105 @@
+}
+
+export class S3Handler implements WebResourceHandler {
+	private readonly bucket: string;
+	private readonly region: string;
+	private readonly accessKeyId: string;
+	private readonly secretAccessKey: string;
+	private readonly endpoint: string;
```
isn't this actually an instance of the S3HandlerConfig?

otaviojacobi replied:
Yep, refactored most of this class.

fisehara commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,105 @@
+	private client: S3Client;
+
+	constructor(config?: S3HandlerConfig) {
+		this.bucket =
+			config?.bucket ??
+			optionalVar('S3_STORAGE_ADAPTER_BUCKET', 'balena-pine-web-resources');
+		this.region =
+			config?.region ?? optionalVar('S3_STORAGE_ADAPTER_REGION', 'us-east-1');
```
Both also don't have default vars as neither the bucket nor the region can be defaulted.
If they are not specified in the config nor in the env the startup should fail

fisehara commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,105 @@
+	private readonly endpoint: string;
+	private client: S3Client;
+
+	constructor(config?: S3HandlerConfig) {
```
I'd either make it a definitive either S3HandlerConfig passed on the constructor or tkae all from env vars.
So that the S3HandlerConfig becomes strongly typed or is undefined and if undefied it gets the details from the env.
Mixing both may be confusing?

otaviojacobi replied:
done

fisehara commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,105 @@
+		const key = `${resource.fieldname}_${uuidv4()}_${resource.originalname}`;
+		const params = {
+			Bucket: this.bucket,
+			ACL: 'public-read',
+			StorageClass: 'STANDARD',
```
Is this a configuration or a fixed usage of the bucker params? If this is the only way the file to bucket works, the bucket needs to be configured this way, right? Please add this in the readme then.

fisehara commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,105 @@
+		await client.send(command);
+	}
+
+	private getS3Client(): S3Client {
+		if (!this.client) {
+			this.client = new S3Client(this.getS3ClientConfig());
+		}
+		return this.client;
+	}
```
Is there any further usage of this method, rather then initialising it once in the constructor? Is there a benefit in this.getS3Client() rather than this.client? Can the client be deleted and needs reintilisation?

otaviojacobi replied:
fixed

fisehara commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,105 @@
+	private getS3URL(key: string) {
+		return this.endpoint
+			? `${this.endpoint}/${this.bucket}/${key}`
+			: `https://${this.bucket}.s3.${this.region}.amazonaws.com/${key}`;
```
I'd rather go with a different way to default this full URL was it may change in syntax or naming convention or ...
I wonder if pinejs even shoul have such a hard coded dependency on amazonaws

otaviojacobi replied:
@fisehara  I don't really know how to proceed on this. I think it is very unlikely aws will change their S3 url pattern as this would be a major problem, at the same time this is only used if you pass a S3 configuration, so you can just decide not to use ....

fisehara reviewed (commented):
-

fisehara commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,327 @@
+	handler: WebResourceHandler,
+	apiRoot?: string,
+) => {
+	if (apiRoot) {
```
Please add an explicit check what apiRoot should be. For later readability it's clearer to know what actual value it should have.

fisehara commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,327 @@
+		},
+		POSTRUN: ({ tx, request }) => {
+			tx.on('end', () => {
+				const keysToDelete: string[] =
+					request.custom.$pineWebResourcesToDelete || [];
+				// on purpose does not await for this promise to resolve
+				deleteFiles(keysToDelete, webResourceHandler);
```
So if the `pineWebResourcesToDelete` is empty the deleteFiles doesn't need to be called, right? I can see that the deleteFiles won't do anything. I'm curious what is cleaner, calling it with empty array or not calling it at all.

otaviojacobi replied:
@fisehara I added as is because I thought having `keysToDelete` / `$pineWebResourcesToDelete` as a `string[]` was cleaner than a `string[] | null` and then having to always check for the null value. But I can change if you think allowing it to be null is better

fisehara commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,327 @@
+
+// TODO: this can be cached
+const getWebResourceFields = (request: uriParser.ODataRequest): string[] => {
+	if (request.abstractSqlModel?.tables[request.resourceName]?.fields == null) {
+		return [];
+	}
+
+	return request.abstractSqlModel?.tables[request.resourceName].fields
+		.filter((f) => f.dataType === 'WebResource')
+		.map((f) => sqlNameToODataName(f.fieldName));
```
Is is potentially not translation save as the tables array may contian: `resourceName`, `resourceName$v1`, `resourceName$latest` resourceNames.
I wonder how to make it translation save, we need to discuss this.

fisehara reviewed (commented):
I wonder if and how to make this translation safe.

otaviojacobi commented on package.json:
```diff
@@ -58,16 +59,21 @@
   },
   "devDependencies": {
     "@balena/lint": "^6.2.1",
+    "@types/busboy": "^1.5.0",
```
@fisehara the @types/busboy is just dev dependency for the typescript typing, but the actual busboy dependency is above under "dependencies"

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,327 @@
+	try {
+		handler = new S3Handler();
+	} catch (e) {
+		console.warn(`Failed to initialize S3 handler, using noop ${e}`);
```
@fisehara My understanding here is that S3 dependency/WebResources is an optional feature of Pinejs. And a backwards compactible one. Let's say you run a pine application which will not use WebResources. You might not want to setup a minio/s3 bucket at all. If ever comes a time where you need it, you add the S3 env variables and it will use S3 handler.

fisehara replied:
@otaviojacobi 
Yeah that's okay but if you have defined it, it should load and not fall back to a noop handler. If you don't need it it should not initialise anything, right?

otaviojacobi replied:
@fisehara  The problem is: Previously, we used multer to handle multi-part requests (either they constained a file or not). Now, we are handling multi-part requests ourselves with Busboy library. To keep the same behaviour as before, even if a file is passed and ignored, busboy requires you to at least close the handle (if you look into the Noop, it just does that: ignores the file), maybe renaming into IgnoreFile handler? Because it is just there to keep multi-part requests support (even if they don't have files)

otaviojacobi commented on package.json:
```diff
@@ -58,16 +59,21 @@
     "pinejs-client-core": "^6.12.3",
     "randomstring": "^1.2.3",
-    "typed-error": "^3.2.1"
+    "type-is": "^1.6.18",
```
This is basically to keep the behaviour from multer https://github.com/expressjs/multer/blob/master/lib/make-middleware.js#L1 but we can discuss if it is worth implementing it ourselves

otaviojacobi commented on src/config-loader/config-loader.ts:
```diff
@@ -196,7 +203,20 @@ export const setup = (app: Express.Application) => {
 						app.options(apiRoute, (_req, res) => res.status(200).end());
-						app.all(apiRoute, sbvrUtils.handleODataRequest);
+
+						const webResourceHandler =
+							model.webResourceHandler ?? getDefaultHandler();
+
+						const fileUploadMiddleware =
+							getUploaderMiddlware(webResourceHandler);
+
+						app.all(
+							apiRoute,
+							fileUploadMiddleware,
+							sbvrUtils.handleODataRequest,
+						);
+
+						setupUploadHooks(webResourceHandler, model.apiRoot);
```
Afaik and tested, the ordering does not change anything

otaviojacobi commented on test/04-webresource.test.ts:
```diff
@@ -0,0 +1,360 @@
+			await expectNotToExist(fileKey);
+		});
+
+		it('fails do update multiple rows with same webresource', async () => {
+			const newFilePath =
+				'test/fixtures/04-webresource/resources/other-image.png';
+			const newFileName = 'other-image.png';
+
+			const res = await supertest(testLocalServer)
+				.patch(`/example/organization`)
+				.attach('logo_image', newFilePath, {
+					filename: newFileName,
+					contentType,
+				})
+				.expect(400);
+
+			expect(res.body)
+				.to.be.a('string')
+				.that.equals(
+					'WebResources can only be updated when providing a resource key.',
+				);
+		});
```
Yes!

otaviojacobi commented on package.json:
```diff
@@ -18,20 +18,20 @@
-    "test": "npm run lint && npm run build && npm run webpack-build && npm run test:compose",
-    "test:compose": "trap 'docker-compose -f docker-compose.npm-test.yml down ; echo Stopped ; exit 0' SIGINT; docker-compose -f docker-compose.npm-test.yml up -d && sleep 2 && DATABASE_URL=postgres://docker:docker@localhost:5431/postgres npm run mocha",
+    "test": "npm run lint & npm run build && npm run webpack-build && npm run test:compose",
+    "test:compose": "docker ps; trap 'docker-compose -f docker-compose.npm-test.yml down ; echo Stopped ; exit 0' INT; docker-compose -f docker-compose.npm-test.yml up -d && sleep 2 && DATABASE_URL=postgres://docker:docker@localhost:5431/postgres S3_ENDPOINT=http://localhost:43680 S3_ACCESS_KEY=AKIAIOSFODNN7EXAMPLE S3_SECRET_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY npm run mocha",
```
Agreed. Removed.

otaviojacobi commented on docker-compose.npm-test.yml:
```diff
@@ -8,4 +8,25 @@ services:
+    restart: always
+    environment:
+      MINIO_ROOT_USER: AKIAIOSFODNN7EXAMPLE
+      MINIO_ROOT_PASSWORD: wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
```
Agreed. Done.

otaviojacobi commented on test/04-webresource.test.ts:
```diff
@@ -0,0 +1,360 @@
+): Promise<string[]> => {
+	const endpoint = `${process.env.S3_ENDPOINT}/${bucket}`;
+	const s3client = new S3Client({
+		region: 'us-east-1',
```
Nope, it just needs the property not to be empty. Used a dummy value for testing.

otaviojacobi commented on test/04-webresource.test.ts:
```diff
@@ -0,0 +1,360 @@
+	});
+
+	after(async () => {
+		await testDeInit(pineServer);
```
Done.

otaviojacobi commented on test/04-webresource.test.ts:
```diff
@@ -0,0 +1,360 @@
+		it('check /ping route is OK', async () => {
+			await supertest(testLocalServer).get('/ping').expect(200, 'OK');
+		});
+	});
```
Removed. Done.

otaviojacobi commented on test/04-webresource.test.ts:
```diff
@@ -0,0 +1,360 @@
+				.attach('logo_image', filePath, { filename, contentType })
+				.expect(201);
+
+			const organization = postRes.body;
```
Done for this one and all the other occurences.

otaviojacobi commented on test/04-webresource.test.ts:
```diff
@@ -0,0 +1,360 @@
+
+	describe('webresource', () => {
+		it('creates a organization with a logo image', async () => {
+			const postRes = await supertest(testLocalServer)
```
Done.

otaviojacobi commented on test/04-webresource.test.ts:
```diff
@@ -0,0 +1,360 @@
+			await expectNotToExist(fileKey);
+		});
+
+		it('creates a organization with a logo image', async () => {
```
Done

otaviojacobi commented on src/sbvr-api/sbvr-utils.ts:
```diff
@@ -1326,8 +1330,13 @@ const runODataRequest = (req: Express.Request, vocabulary: string) => {
 };
 
-export const handleODataRequest: Express.Handler = async (req, res, next) => {
+export const getApiRoot = (req: Express.Request): string | null => {
```
Done.

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,327 @@
+	handler: WebResourceHandler,
+	apiRoot?: string,
+) => {
+	if (apiRoot) {
```
Done

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,327 @@
+
+export const setupUploadHooks = (
+	handler: WebResourceHandler,
+	apiRoot?: string,
```
added apiRoot as mandatory parameter

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,327 @@
+
+const getWebResourcesHrefs = (webResources?: WebResourcesDbResponse[]) => {
+	const resources = webResources ?? [];
+	const hrefs = resources
```
nice one! done.

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,327 @@
+	removeFile: (fileReference: string) => Promise<void>;
+}
+
+type WebResourcesDbResponse =
```
@fisehara it is for the case for something like:
```js
{
    "propertyName": null
}
```

otaviojacobi commented on test/04-webresource.test.ts:
```diff
@@ -0,0 +1,360 @@
+			await expectNotToExist(uniqueFilename);
+		});
+
+		it('deletes file if transaction type rule is not respected', async () => {
```
done!

otaviojacobi commented on test/04-webresource.test.ts:
```diff
@@ -0,0 +1,360 @@
+			await expectNotToExist(uniqueFilename);
+		});
+
+		it('ignores files in not post or patch requests', async () => {
```
Added for delete and options

otaviojacobi commented on test/04-webresource.test.ts:
```diff
@@ -0,0 +1,360 @@
+
+			await expectNotToExist(uniqueFilename);
+		});
+	});
```
Done! I also added a comment about this. Currently if the file is deleted in s3 (e.g. manually) we will still serve the dangling reference. But added a comment we might want to change it in the future

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,327 @@
+						},
+					},
+				},
+			})) as WebResourcesDbResponse[] | undefined;
```
done!

fisehara commented on test/05-webresource.test.ts:
```diff
@@ -0,0 +1,472 @@
+				// Create a large file by using an image at the head ( so that filetype can identify it as an image ),
+				// and then appending a large chunk
+				const { largeStream, largeFileSize } = await getLargeFileStream(
+					1024 * 1024 * 512 - fileSize,
```
Shouldn't this file size fail?

otaviojacobi replied:
This creates a filestream that is slightly smaller than the size on the SBVR.

fisehara commented on package.json:
```diff
@@ -18,20 +18,20 @@
-    "test": "npm run lint && npm run build && npm run webpack-build && npm run test:compose",
-    "test:compose": "trap 'docker-compose -f docker-compose.npm-test.yml down ; echo Stopped ; exit 0' SIGINT; docker-compose -f docker-compose.npm-test.yml up -d && sleep 2 && DATABASE_URL=postgres://docker:docker@localhost:5431/postgres npm run mocha",
+    "test": "npm run lint & npm run build && npm run webpack-build && npm run test:compose",
+    "test:compose": "trap 'docker-compose -f docker-compose.npm-test.yml down ; echo Stopped ; exit 0' SIGINT; docker-compose -f docker-compose.npm-test.yml up -d && sleep 2 && DATABASE_URL=postgres://docker:docker@localhost:5431/postgres PINEJS_WEBRESOURCE_MAXFILESIZE=1000000000 S3_ENDPOINT=http://localhost:43680 S3_ACCESS_KEY=USERNAME S3_SECRET_KEY=PASSWORD npm run mocha",
```
Is there a value in testing for 1GiB rather than couple of MiBs?

otaviojacobi replied:
We need a way to validate we can upload large files (500MiB for example). But if we want to remove the test for uploading large files (which tbh I don't think we should), then we can lower this down.

fisehara replied:
Yeah I agree we should keep the large file tests.

fisehara reviewed (commented):
Some NITs, looking good some questions.

otaviojacobi commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,99 @@
+		} catch (err: any) {
+			if (err?.name === 'AbortError') {
+				resource.stream.resume();
+				throw new Error('File size exceeded');
```
I see what you mean. I think we can use the closure here and change the check to be only regarding the size:
```ts
resource.stream.resume();
if (size > this.maxFileSize) { 
    throw new Error('File size exceeded');
}
throw error;
```

otaviojacobi replied:
done

otaviojacobi commented on test/05-webresource.test.ts:
```diff
@@ -0,0 +1,472 @@
+					})
+					.expect(400);
+
+				expect(res.body).to.include('It is necessary that');
```
@fisehara My idea here is just to have a way to differentiate the 400 resulted from the rules vs the 400 resulting from a file exceeding the environment variable, do you have any other idea for this?

fisehara replied:
@otaviojacobi This differentiation is fine, my thought is about maintainability / readability. When I run the tests and see that it should return `It is necessary that` but doesn't, where do I search for this. That's why I meant it should have the full rule string to check, so that it's clear where to find the cause for this test.

otaviojacobi commented on test/05-webresource.test.ts:
```diff
@@ -0,0 +1,472 @@
+				await expectNotToExist(fileKey);
+			});
+
+			it(`fails to update multiple rows with same ${resourcePath}`, async () => {
+				const newFilePath =
+					'test/fixtures/05-webresource/resources/other-image.png';
+				const newFileName = 'other-image.png';
+
+				const res = await supertest(testLocalServer)
+					.patch(`/${resourceName}/organization`)
+					.attach(resourcePath, newFilePath, {
+						filename: newFileName,
+						contentType,
+					})
+					.expect(400);
+
+				expect(res.body)
+					.to.be.a('string')
+					.that.equals(
+						'WebResources can only be updated when providing a resource key.',
+					);
+			});
```
That's a good question. I don't know. I think initially we want it to be a simple 1-0/1 relationship.

fisehara commented on test/05-webresource.test.ts:
```diff
@@ -0,0 +1,472 @@
+
+			it(`deletes ${resourcePath} if transaction size rule is not respected`, async () => {
+				const uniqueFilename = `${randomUUID()}_${filename}`;
+				const { largeStream } = await getLargeFileStream(
+					1024 * 1024 * 600,
+					filePath,
+				);
```
@otaviojacobi just test it with 1GiB + 1 file size then. I'm fine with the large file tests.

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,325 @@
+			const { api, request } = args;
+			let fields = getWebResourceFields(request);
+
+			if (fields.length === 0) {
+				return;
+			}
```
@fisehara you mean should not be too expensive?

fisehara replied:
Yeah, it's not too expensive that justifies checking `fields.length === 0` twice?

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,325 @@
+				return;
+			}
+
+			const ids = await sbvrUtils.getAffectedIds(args);
+			if (ids.length === 0) {
+				return;
+			}
+
+			if (ids.length !== 1) {
+				throw new errors.BadRequestError(
+					'Resources containing webresources can only be updated/deleted one at a time.',
+				);
+			}
```
Nice one!

otaviojacobi replied:
done

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,325 @@
+		'PATCH',
+		apiRoot,
+		resourceName,
+		getRemoveWebResourceHook(handler),
```
@fisehara  You mean the function name? from getRemoveWebResourceHook to something like `getRemoveWebResourceHooksOnPatchAndDelete` ?

fisehara replied:
Hmm, after thinking about it I understand that it's correctly named as it removes the WebResource. Although the usage seems counter-intuitive to be used in the Patch as well. Maybe outline this with a comment on the patch hook.
Anyway, it's a NIT.

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,325 @@
+				options: {
+					$select: fields,
+				},
+			})) as WebResourcesDbResponse | undefined | null;
```
done

otaviojacobi commented on test/05-webresource.test.ts:
```diff
@@ -0,0 +1,472 @@
+		region: 'dummy-region',
+		credentials: {
+			accessKeyId: process.env.S3_ACCESS_KEY as string,
+			secretAccessKey: process.env.S3_SECRET_KEY as string,
```
done

otaviojacobi commented on test/05-webresource.test.ts:
```diff
@@ -0,0 +1,472 @@
+	// Inspects minio bucket to ensure no file with this uuid_filename exists
+
+	const files = await listAllFilesInBucket(
+		process.env.S3_STORAGE_ADAPTER_BUCKET,
```
done

otaviojacobi commented on test/05-webresource.test.ts:
```diff
@@ -0,0 +1,472 @@
+
+			it('does not fail to serve if S3 resource is deleted but entry exists', async () => {
+				// This tets the current behaviour, but we might want to change it in the future
+				// because the current behaviour allows for a dangling reference to exist
```
done

otaviojacobi commented on test/05-webresource.test.ts:
```diff
@@ -0,0 +1,472 @@
+	// Inspects minio bucket to ensure no file with this uuid_filename exists
+
+	const files = await listAllFilesInBucket(
+		process.env.S3_STORAGE_ADAPTER_BUCKET,
```
done

fisehara commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,327 @@
+			} catch (err: any) {
+				await clearFiles();
+
+				if (err?.message.includes('File size exceeded')) {
+					return sbvrUtils.handleHttpErrors(
+						req,
+						res,
+						new errors.BadRequestError('File size exceeded'),
+					);
+				}
```
Or more specific WebResourceError

otaviojacobi replied:
@fisehara  I guess FileSizeExceededError no? Because that's the only case where we want to throw the 400. The others are really 500. Maybe we can have both where FileSizeExceededError extends WebResourceError

fisehara replied:
@otaviojacobi that sounds appropriate, yes!

fisehara commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,99 @@
+			await upload.done();
+		} catch (err: any) {
+			resource.stream.resume();
+			if (size > this.maxFileSize) {
+				throw new Error('File size exceeded');
+			}
```
Or more specific WebResourceError

Page- commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,344 @@
+			let fields = getWebResourceFields(request);
+
+			if (request.method === 'PATCH') {
+				if (request.odataQuery?.key == null) {
```
Shouldn't this check come after we have ensured that at least one field matches a webresource field?

otaviojacobi replied:
@Page- My understanding is that no. This is a limitation for fileuploads as it must have a odata key (e.g. you can't upload the same file to multiple resources at the same time). So it doesn't really matter who comes first here

Page- commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,344 @@
+
+const isDefined = <T>(x: T | undefined | null): x is T => x != null;
+
+const getWebResourcesHrefs = (webResources?: WebResourcesDbResponse | null) => {
```
I'd appreciate it if you could add the response type here

Page- commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,344 @@
+				}
+
+				const keysToDelete: string[] = fields
+					.filter((f) => isDefined(request.values[f]))
+					.map((f) => request.values[f].href);
```
I think this might flow better and have less lookups as:
```suggestion
					.map((f) => request.values[f]?.href);
					.filter(isDefined)
```

otaviojacobi replied:
@Page- That is a very nice catch, I think it was automatic to first filter and then map, map in this very specific case we can have less lookups as previously both operations have to do the request.values[f] access. Thanks!

Page- commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,344 @@
+const getModifiedFields = (request: uriParser.ODataRequest): string[] => {
+	return Object.entries(request.values)
+		.filter(([_key, value]) => value !== undefined)
+		.map(([key, _value]) => key);
```
```suggestion
		.map(([key]) => key);
```

Page- commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,344 @@
+};
+
+// TODO: this can be cached
+const getWebResourceFields = (request: uriParser.ODataRequest): string[] => {
```
If you do some destructuring here it'll make the rest a lot shorter, and also clarify which specific properties are important
```suggestion
const getWebResourceFields = ({ abstractSqlModel, resourceName }: uriParser.ODataRequest): string[] => {
```

otaviojacobi replied:
Thanks, I did it for this function and a few others

Page- commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,344 @@
+			isAborting = true;
+			const deletions = filesUploaded.map((file) => handler.removeFile(file));
+			// Best effort: We try to remove all uploaded files, but if this fails, there is not much to do
+			return Promise.all(deletions).catch((err) =>
+				console.error('Error deleting file', err),
+			);
```
```suggestion
			return Promise.all(deletions).catch((err) =>
				console.error('Error deleting file', err);
			});
```

otaviojacobi replied:
thanks! done

Page- commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,344 @@
+	originalError?: Error;
+}
+
+export class FileSizeExceededError extends WebResourceError {
+	constructor(maxSize: number) {
+		super(`File size exceeded the limit of ${maxSize} bytes.`);
+		this.name = 'FileSizeExceededError';
+	}
+}
```
```suggestion
export class FileSizeExceededError extends WebResourceError {
	name = 'FileSizeExceededError';
	constructor(maxSize: number) {
		super(`File size exceeded the limit of ${maxSize} bytes.`);
	}
}
```

otaviojacobi replied:
done.

Page- commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,344 @@
+	removeFile: (fileReference: string) => Promise<void>;
+}
+
+export class WebResourceError extends Error {
+	constructor(message: string, originalError?: Error) {
+		super(message);
+		this.name = 'WebResourceError';
+
```
```suggestion
export class WebResourceError extends Error {
	name = 'WebResourceError';
	constructor(message: string, originalError?: Error) {
		super(message);

```

fisehara commented on test/06-webresource.test.ts:
```diff
@@ -0,0 +1,501 @@
+							filename,
+							contentType: 'text/csv',
+						})
+						.expect(400);
```
```suggestion
						.expect(400);
					fileKey = ....
					await expectNotToExist(fileKey);
```

fisehara reviewed (commented):
Overall looks good to me. Needs some added comments to increase read / maintainability.

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,355 @@
+		return false;
+	}
+
+	// TODO: This could be cached
```
@fisehara Actually no. All the todos left were things I thought while coding this "If we run into performance issues, this is where it could be improved". My understanding is that a lot of caching is already going on in the background and adding more on top of it just for the sake of "it might be a performance improvement" would make it even harder to mantain. But I also think that lots of these TODOs around look ugly/incomplete. What do you think of removing them from the code and storing as tips somewhere in case we need to revisit/improve performance?

otaviojacobi commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,98 @@
+	constructor() {
+		this.endpoint = requiredVar('S3_ENDPOINT');
+		this.config = {
+			region: optionalVar('S3_REGION', 'us-east-1'),
```
After having some pain understanding key differences between AWS S3 and Minio S3 I agree. This should be a required var

otaviojacobi replied:
done

otaviojacobi commented on src/server-glue/webresource-handlers/S3Handler.ts:
```diff
@@ -0,0 +1,98 @@
+	private client: S3Client;
+
+	constructor() {
+		this.endpoint = requiredVar('S3_ENDPOINT');
```
Much better using `this.config.endpoint`, thanks!

otaviojacobi replied:
done

otaviojacobi commented on test/06-webresource.test.ts:
```diff
@@ -0,0 +1,501 @@
+const pipeline = util.promisify(pipelineRaw);
+const fs = fsBase.promises;
+
+describe('05 webresources tests', function () {
```
done

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,355 @@
+		if (!is(req, ['multipart'])) {
+			return next();
+		}
+		const filesUploaded: string[] = [];
```
Renamed into `uploadedFilePaths`

otaviojacobi replied:
done

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,355 @@
+			}
+		});
+
+		bb.on('field', (name, val, _info) => {
```
Sure. I added this: 
```
// multipart requests will have two main parts, the file contents and the form fields
// This receives the form fields and transforms them into a standard JSON body
// This is a similar behavior as previous multer library did
```

otaviojacobi replied:
done

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,355 @@
+		}
+	}
+
+	// TODO: We could do a pre-check if there is a SBVR rule specifying file max size
```
removed

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,355 @@
+			const deletions = filesUploaded.map((file) => handler.removeFile(file));
+			// Best effort: We try to remove all uploaded files, but if this fails, there is not much to do
+			return Promise.all(deletions).catch((err) => {
+				console.error('Error deleting file', err);
```
done

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,355 @@
+					);
+				}
+
+				console.error('Error uploading file', err);
```
done

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,355 @@
+	};
+};
+
+// TODO: this can be cached
```
removed

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,355 @@
+
+		bb.on('error', async (err) => {
+			await clearFiles();
+			done();
```
Renamed into `finishFileUpload`

fisehara commented:
@otaviojacobi Lookings good from my side now. What's blocking to elevate from draft?

otaviojacobi commented on docker-compose.npm-test.yml:
```diff
@@ -27,6 +27,5 @@ services:
       /bin/sh -c "
       /usr/bin/mc config host add minio-server http://minio-server:9000 USERNAME PASSWORD;
       /usr/bin/mc mb --ignore-existing minio-server/balena-pine-web-resources;
-      /usr/bin/mc anonymous set download minio-server/balena-pine-web-resources;
```
Odly enough, this is the most important line in this commit. Now the bucket is private and files within it are only accessible if you have a signed url.

fisehara reviewed (commented):
Needs the sbvr-types to be updated to `Content Type` I guess. 

Other than this just NITs and hint for passing the webresource as config rather than model property.
:+1:

otaviojacobi commented on test/fixtures/06-webresource/example.sbvr:
```diff
@@ -0,0 +1,32 @@
+Fact Type: organization has name
+	Necessity: each organization has exactly one name
+
+Fact Type: organization has shortcode
+	Necessity: each organization has at most one shortcode
+	Necessity: each organization that has a shortcode, has a shortcode that has a Length (Type) that is greater than 0 and is less than or equal to 5
```
Agree

otaviojacobi commented on test/06-webresource.test.ts:
```diff
@@ -0,0 +1,743 @@
+						})
+						.expect(201);
+
+					const { body: res } = await supertest(testLocalServer)
+						.patch(`/${resourceName}/organization(${org.id})`)
+						.field('shortcode', '123456')
+						.attach(resourcePath, newFilePath, {
+							filename: otherUniqueFilename,
+							contentType,
+						})
+						.expect(400);
+
+					expect(res).to.be.eq(
+						'It is necessary that each organization that has a shortcode, has a shortcode that has a Length (Type) that is greater than 0 and is less than or equal to 5',
+					);
```
Much better indeed

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,436 @@
+	[fieldname: string]: { href: string } | undefined | null;
+};
+
+const getLogger = (vocab?: string): Console => {
+	if (vocab) {
+		return sbvrUtils.api[vocab]?.logger ?? console;
+	}
+	return console;
+};
```
If you don't mind as this is a very specific minor change and this PR is already incredibly bloated I can do this in patch/minor on top

fisehara replied:
@otaviojacobi that's why I marked it NIT :) It's fine and solid

otaviojacobi commented on test/06-webresource.test.ts:
```diff
@@ -0,0 +1,743 @@
+	);
+});
+
+const removesSigning = (href: string): string => {
+	return href.split('?')[0];
+};
```
After thinking it through @fisehara I don't think it makes sense to test this as this is straight up related on how the provider handles signing (e.g. if/when AWS changes its signing method we shouldn't care, as long as their provider can still verify it is a valid key in the whatever format they want). I think the best test here is to assert we can access the resource with the keys and not access it without it

otaviojacobi replied:
done

otaviojacobi commented:
> Needs the sbvr-types to be updated to `Content Type` I guess.
> 
> Other than this just NITs and hint for passing the webresource as config rather than model property. 👍

@fisehara From my perspective this PR is done. I added several more cases and tested it here https://github.com/balena-io/balena-api/pull/4539 I am available for any needed improvement if you find any and I think it is quite straight forward to rollback if needed (at least just bumping without adding any change to the model won't cause any difference in the API as it is today) but I think it won't be necessary. The next steps are basically having it promoted to the API and then set up the bucket/cloudfront for staging and bucket needed for balenaMachine

otaviojacobi commented on package.json:
```diff
@@ -71,7 +74,8 @@
     "@types/terser-webpack-plugin": "^5.2.0",
-    "@types/webpack": "^5.28.1",
+    "@types/type-is": "^1.6.3",
+    "@types/webpack": "^5.28.0",
```
Very good catch, I guess this is some merge conflict, I will rebase and remove any downgrade Thanks

otaviojacobi commented on src/config-loader/config-loader.ts:
```diff
@@ -195,15 +202,37 @@ export const setup = (app: Express.Application) => {
 							.put(sbvrUtils.handleODataRequest)
-							.post(sbvrUtils.handleODataRequest)
-							.patch(sbvrUtils.handleODataRequest)
+							.post(fileUploadMiddleware, sbvrUtils.handleODataRequest)
```
@Page- We can add but I think we would be better off doing this in a follow up PR if you don't mind. Mostly because POST/PATCH are the ones that will be consumed right now

otaviojacobi commented on package.json:
```diff
@@ -51,18 +51,21 @@
     "randomstring": "^1.2.3",
-    "typed-error": "^3.2.2"
+    "type-is": "^1.6.18",
+    "typed-error": "^3.2.1"
```
Fixed

otaviojacobi commented on src/config-loader/config-loader.ts:
```diff
@@ -195,15 +202,37 @@ export const setup = (app: Express.Application) => {
+							.patch(fileUploadMiddleware, sbvrUtils.handleODataRequest)
 							.merge(sbvrUtils.handleODataRequest)
 							.delete(sbvrUtils.handleODataRequest);
+						app.options(apiRoute, (_req, res) => res.status(200).end());
```
removed

otaviojacobi commented on src/config-loader/config-loader.ts:
```diff
@@ -195,15 +202,37 @@ export const setup = (app: Express.Application) => {
+							);
+
+							const resourceNames = resources.map(
+								([r, _table]) => loadedModel.abstractSql.tables[r].name,
```
Indeed this whoel block can be translated into:
```ts
const resourceNames = Object.values(
loadedModel.abstractSql.tables,
).filter((table) =>
table.fields.some((f) => f.dataType === 'WebResource'),
).map((table) => table.name);
```

otaviojacobi replied:
done

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,434 @@
+	onPreRespond: (webResource: WebResource) => Promise<WebResource>;
+}
+
+export class WebResourceError extends Error {
```
Indeed this package is much better :)

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,434 @@
+				response.body.d = transformedItems;
+			}
+		},
+	} as sbvrUtils.Hooks;
```
For some reason I had done this for `getRemoveWebResourceHooks` and skipped for the others. Fixed now

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,434 @@
+			request.custom.$pineWebResourcesToDelete =
+				getWebResourcesHrefs(webResources);
+		},
+		POSTRUN: ({ tx, request }) => {
+			// Successfully patched or deleted resources
+			tx.on('end', () => {
+				deleteRequestPendingFiles(request, webResourceHandler);
+			});
+		},
+		'POSTRUN-ERROR': ({ request, tx }) => {
+			// Request failed before DB roundtrip (e.g. Trying to PATCH multiple with single webresource)
+			// Or by any API rule (e.g. allowlist)
+			tx?.on('end', () => {
+				deleteRequestPendingFiles(request, webResourceHandler);
+			});
+
+			// Request failed on DB roundtrip (e.g. DB constraint)
+			tx?.on('rollback', () => {
+				deleteRollbackPendingFields(request, webResourceHandler);
+			});
+		},
```
@Page- I don't 100% follow here. If we add these two `tx.on` in the PRERUN we don't need the POSTRUN/POSTRUN-ERROR then?

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,434 @@
+			// Successfully patched or deleted resources
+			tx.on('end', () => {
+				deleteRequestPendingFiles(request, webResourceHandler);
+			});
```
done

otaviojacobi commented:
OK @Page- because changing of the folder structure some of the conversation is a bit lost so I will summarize here:
I addressed all your comments but two: 
1) Presumably this should also be added to PUT and MERGE?
Yes, it can, but if you are ok with it I would prefer to have it done in a separate PR not to add more stuff in here

2) It probably makes sense to add these cleanup hooks as part of the PRERUN instead
The reason I didn't put all in the PRERUN hook is because the PRERUN hook can itself throw. If it does throw I need to have the keysToDelete before hand which would endup adding the tx.on('rollback') in two separate paths in the code. Maybe I could move the tx.on('end') to the PRERUN then?

otaviojacobi commented on src/server-glue/webresource-handler.ts:
```diff
@@ -0,0 +1,434 @@
+
+			if (typeof response.body === 'object' && response.body.d != null) {
+				const transformedItems = await Promise.all(
+					response.body.d.map(async (responseItem: any) => {
```
@Page-  I don't see exactly how as this comes from: https://github.com/balena-io/pinejs/blob/master/src/sbvr-api/sbvr-utils.ts#L142

In the future I want to refactor this to somehow call the `webResourceHandler.onPreRespond` at the momento the content is actually fetched from the DB (maybe another hook?) And then possibly this any might disappear and we get all the `$expand` also working for this (right now it just works if you $expand on the model that has the webresource). But I think this should can be an isolated improvement on its own

fisehara commented on src/webresource-handler/index.ts:
```diff
@@ -0,0 +1,426 @@
+): void => {
+	const keysToDelete: string[] = request.custom.$pineWebResourcesToDelete || [];
+	// on purpose does not await for this promise to resolve
+	deleteFiles(keysToDelete, webResourceHandler).catch((err) => {
+		getLogger(request.vocabulary).error(`Failed to delete pending files`, err);
+	});
```
```suggestion
    try {
	    deleteFiles(keysToDelete, webResourceHandler);
    } catch (err) {
        getLogger(request.vocabulary).error(`Failed to delete pending files`, err);
    }
```

fisehara reviewed (commented):
NITs

I've tried to move the tx.on('end') and tx.on('rollback") into the prerun, but the tests are failling.
Basically I understand the idea, as the `tx` should be the same through the different execution states.

otaviojacobi commented:
@Page- @fisehara Ok, so the last commit adds all the deletion flow in the PRERUN, it is not the most elegant thing (as tx.on('end') hook is set conditionally depending on the request, but it works and avoids extra hooks). The only comment left is regarding MERGE/PUT requests. If this is a blocker for merging this let me know, directly and I can add in here, otherwise I would really like to get this into balena api still this week if possible. I added a separate commit with just the last fixes but I want to squash all commits and don't want to loose the whole context before review

fisehara commented:
@otaviojacobi LGTM squash the commits and I'll approve it.
